File: dart/dna424_client/test/attest_client_test.dart
import 'dart:convert';

import 'package:dna424_client/dna424_client.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:test/test.dart';

/// AttestClient HTTP behaviour: non-2xx responses throw AttestApiError, and
/// revokeIssuer hits the right route. Mirrors the TS SDK's attestClient tests.
void main() {
  AttestClient clientReturning(int status, Map<String, dynamic> body,
      {void Function(http.Request)? onRequest}) {
    final mock = MockClient((req) async {
      onRequest?.call(req);
      return http.Response(jsonEncode(body), status,
          headers: {'content-type': 'application/json'});
    });
    return AttestClient('https://x', client: mock);
  }

  test('verify returns the parsed result on 200', () async {
    final c = clientReturning(200, {
      'chip_authentic': true,
      'not_replayed': true,
      'issuer_verified': true,
      'reason': '',
    });
    final r = await c.verify('04D2', 1, 'aa');
    expect(r.fullyValid, isTrue);
  });

  test('non-2xx throws AttestApiError with status + detail', () async {
    final c = clientReturning(400, {'detail': 'm is not valid hex'});
    expect(
      () => c.verify('04D2', 1, 'ZZ'),
      throwsA(isA<AttestApiError>()
          .having((e) => e.status, 'status', 400)
          .having((e) => e.detail, 'detail', contains('not valid hex'))),
    );
  });

  test('registerIssuer returns issuer_id and token', () async {
    final c = clientReturning(200, {'issuer_id': 'iss_abc', 'token': 'isk_x'});
    final reg = await c.registerIssuer('04ab');
    expect(reg.issuerId, equals('iss_abc'));
    expect(reg.token, equals('isk_x'));
  });

  test('revokeIssuer POSTs the revoke route', () async {
    http.Request? seen;
    final c = clientReturning(200, {'issuer_id': 'iss_abc', 'status': 'revoked'},
        onRequest: (r) => seen = r);
    await c.revokeIssuer('iss_abc');
    expect(seen!.method, equals('POST'));
    expect(seen!.url.path, equals('/v1/issuers/iss_abc/revoke'));
  });

  test('revokeIssuer throws on 403', () async {
    final c = clientReturning(403, {'detail': 'cannot revoke another issuer'});
    expect(() => c.revokeIssuer('iss_other'), throwsA(isA<AttestApiError>()));
  });
}

File: ts/dna424-client/test/attestClient.test.mjs
/**
 * AttestClient HTTP behaviour: non-2xx responses throw AttestApiError (not a
 * success body cast to the wrong type), and revokeIssuer hits the right route.
 */
import test from "node:test";
import assert from "node:assert/strict";

import { AttestClient, AttestApiError } from "../dist/index.js";

function mockFetch(handler) {
  const calls = [];
  globalThis.fetch = async (url, init = {}) => {
    calls.push({ url: String(url), method: init.method ?? "GET" });
    const { status = 200, body = {} } = handler(String(url), init) ?? {};
    return new Response(JSON.stringify(body), {
      status,
      headers: { "content-type": "application/json" },
    });
  };
  return calls;
}

test("verify returns the parsed result on 200", async () => {
  mockFetch(() => ({ body: { chip_authentic: true, not_replayed: true, issuer_verified: true, reason: "" } }));
  const c = new AttestClient("https://x");
  const r = await c.verify("04D2", 1, "aa");
  assert.equal(r.chip_authentic, true);
});

test("non-2xx throws AttestApiError with status + detail", async () => {
  mockFetch(() => ({ status: 400, body: { detail: "m is not valid hex" } }));
  const c = new AttestClient("https://x");
  await assert.rejects(c.verify("04D2", 1, "ZZ"), (e) => {
    assert.ok(e instanceof AttestApiError);
    assert.equal(e.status, 400);
    assert.match(e.detail, /not valid hex/);
    return true;
  });
});

test("registerIssuer returns issuer_id and token", async () => {
  mockFetch(() => ({ body: { issuer_id: "iss_abc", token: "isk_x" } }));
  const c = new AttestClient("https://x");
  const reg = await c.registerIssuer("04ab");
  assert.equal(reg.issuerId, "iss_abc");
  assert.equal(reg.token, "isk_x");
});

test("revokeIssuer POSTs the revoke route and resolves", async () => {
  const calls = mockFetch(() => ({ body: { issuer_id: "iss_abc", status: "revoked" } }));
  const c = new AttestClient("https://x", "tok");
  await c.revokeIssuer("iss_abc");
  assert.equal(calls[0].method, "POST");
  assert.match(calls[0].url, /\/v1\/issuers\/iss_abc\/revoke$/);
});

test("revokeIssuer throws on 403", async () => {
  mockFetch(() => ({ status: 403, body: { detail: "cannot revoke another issuer" } }));
  const c = new AttestClient("https://x");
  await assert.rejects(c.revokeIssuer("iss_other"), AttestApiError);
});
